### ISSUE TITLE
Integrating eventing from Neo4j Sessions into the autoconfig.

### ISSUE THREAD

**Issue by mangrish** [spring-projects/spring-boot#7531]
Hi,

This is something between a question and an enhancement request.
I have created a Spring Boot app (1.5.0.BUILD-SNAPSHOT) with the Ingalls BUILD-SNAPSHOT of Spring Data Neo4j (SDN).

The problem I'm facing is that I'm not sure how to allow developers to integrate events originating from the OGM into Spring without having to exclude the whole autoconfiguration and rewiring it all again with the eventing code. The OGM `SessionFactory.openSession()` needs to be overridden to propagate the events into Spring.

This is how I achieved it in Spring Boot 1.4.x (with the same Ingalls BUILD-SNAPSHOT):
https://github.com/neo4j-examples/sdn4-university/blob/4.2.x/src/main/java/school/Application.java
and
https://github.com/neo4j-examples/sdn4-university/blob/4.2.x/src/main/java/school/Neo4jConfiguration.java

The "new" version I'm working on is here: https://github.com/neo4j-examples/sdn4-university/blob/spring-boot-1.5/src/main/java/school/Application.java

I'm hoping I don't have to force developers to exclude the autoconfiguration just to achieve this.

Any ideas on how to proceed?

**Comment by wilkinsona** [comment]
We could do something like this in `Neo4jDataAutoConfiguration`:

```java
@Bean
public SessionFactory sessionFactory(org.neo4j.ogm.config.Configuration configuration,
		ApplicationContext applicationContext,
		final ObjectProvider<List<EventListener>> eventListenersProvider) {
	return new SessionFactory(configuration, getPackagesToScan(applicationContext)) {

		@Override
		public Session openSession() {
			Session session = super.openSession();
			List<EventListener> eventListeners = eventListenersProvider
					.getIfAvailable();
			if (CollectionUtils.isNotEmpty(eventListeners)) {
				for (EventListener eventListener : eventListeners) {
					session.register(eventListener);
				}
			}
			return session;
		}

	};
}
```

This will automatically register any `EventListener` beans with every `Session` that's created by the auto-configured `SessionFactory`. Is that an appropriate thing to do from a Neo4j perspective?

FWIW, I think the Neo4j API could be a little friendlier here. Given that you want auto-configuration to support this use case, I assume that registering an `EventListener` with every `Session` is a common thing to do. Could `SessionFactory` provide first-class support for it by providing an option to configure it with a list of listeners? That would remove the need to subclass it and override `openSession`.

**Comment by mangrish** [comment]
> FWIW, I think the Neo4j API could be a little friendlier here. 

@wilkinsona You know what; that's a complete brain fade on my part. You are right.  This could  just add this to `SessionFactory`. 

Let me hook this up in a branch and see how we end up. I'll circle back to this in a couple of days (or earlier if time permits).

**Comment by spring-issuemaster** [comment]
If you would like us to look at this issue, please provide the requested information. If the information is not provided within the next 7 days this issue will be closed.

**Comment by mangrish** [comment]
@wilkinsona We have added a `register/deregister` to the `SessionFactory` as you suggested: https://github.com/neo4j/neo4j-ogm/blob/master/core/src/main/java/org/neo4j/ogm/session/SessionFactory.java

This now allows event listeners to work across all sessions.

The configuration I'm wiring up is here: https://github.com/neo4j-examples/sdn4-university/blob/spring-boot-1.5/src/main/java/school/Application.java

What's the best way to get the `EventPublisher` bean connected to the `Neo4jDataAutoConfiguration.sessionFactory()`?

**Comment by mangrish** [comment]
I've updated https://github.com/neo4j-examples/sdn4-university/blob/spring-boot-1.5/src/main/java/school/Application.java and added a method straight after Boot initialisation. It's not pretty but it works. Is there a more "Boot"y way of doing this?

**Comment by wilkinsona** [comment]
I think we should update Boot's auto-configuration so that the auto-configured `SessionFactory` automatically has any `EventListener` beans registered with it:

```java
	@Bean
	public SessionFactory sessionFactory(org.neo4j.ogm.config.Configuration configuration,
			ApplicationContext applicationContext,
			ObjectProvider<List<EventListener>> eventListenersProvider) {
		SessionFactory sessionFactory = new SessionFactory(configuration,
				getPackagesToScan(applicationContext));
		List<EventListener> providedEventListeners = eventListenersProvider
				.getIfAvailable();
		if (providedEventListeners != null) {
			for (EventListener eventListener : providedEventListeners) {
				sessionFactory.register(eventListener);
			}
		}
		return sessionFactory;
	}
```

@mangrish What are your plans for another milestone, RC, or release of Neo4J OGM that contains the changes to `SessionFactory`? We're planning Spring Boot 1.5.0.M1 next week so using OGM snapshots isn't really an option at the moment.

**Comment by mangrish** [comment]
@wilkinsona So if i understand this correctly all i need to do is define any amount of `EventListener` `@Bean`s and it will automatically get registered? That is pretty neat!

We released OGM `2.1.0` (GA) over the weekend to coincide with the Neo4j 3.1 release today.
Spring Data Neo4j's next release will be `Ingalls RC1` which pulls in OGM `2.1.0`.

We suspect the OGM will have a patch release of `2.1.1` to coincide with `Ingalls RELEASE`

**Comment by wilkinsona** [comment]
> So if i understand this correctly all i need to do is define any amount of EventListener @Beans and it will automatically get registered?

Correct.

> We released OGM 2.1.0 (GA) over the weekend to coincide with the Neo4j 3.1 release today.
Spring Data Neo4j's next release will be Ingalls RC1 which pulls in OGM 2.1.0.

Excellent, thanks. I missed the 2.1.0 GA as I just checked the releases on GitHub where 2.1.0-M01 seems to be the latest. Looking on Maven Central I can see 2.1.0-RC01 and 2.1.0 are available too.
